src/main/scala/com/evolution/scache/LoadingCache.scala

```diff
@@ -242,91 +312,131 @@ private[scache] object LoadingCache {
                                   .traverse { _.start }
                                   .map { fiber =>
                                     fiber
                                       .foldMapM { _.joinWithNever }
                                       .as { entry0.value.some }
-                                      .asRight[Int]
+                                      .asRight[Int] // Breaking outer cycle
+                                      .asRight[Int] // Breaking inner cycle
                                   }
+                              // Failed to set the entryRef to our value, it is safe to retry
                               case false =>
-                                (counter + 1)
-                                  .asLeft[F[Option[V]]]
+                                (counter1 + 1)
+                                  .asLeft[Either[Int, F[Option[V]]]]
                                   .pure[F]
                             }
 
-                        case (Left(deferred), set) =>
-                          deferred
+                        // The value is still loading, so we first replace it with our value,
+                        // and then try to complete the deferred with it.
+                        case (state: EntryState.Loading[F, V], set) =>
+                          state
+                            .deferred
                             .complete(entry.asRight)
                             .flatMap {
-                              case true  =>
-                                set(entry.asRight).map {
-                                  case true  =>
+                              // We successfully completed the deferred, now trying to set the value.
+                              case true =>
+                                set(EntryState.Value(entry)).flatMap {
+                                  // We successfully replaced the entry with our value, so we are done.
+                                  case true =>
                                     none[V]
                                       .pure[F]
-                                      .asRight[Int]
+                                      .asRight[Int] // Breaking outer cycle
+                                      .asRight[Int] // Breaking inner cycle
+                                      .pure[F]
+                                  // Another fiber completed placed their new value before us
+                                  // so we just release our value and exit.
                                   case false =>
-                                    (counter + 1).asLeft[F[Option[V]]]
+                                    entry
+                                      .release
+                                      .traverse { _.start } // Start releasing and forget
+                                      .as {
+                                        none[V]
+                                          .pure[F]
```

Comment:
`deferred.get` here?

Reply:
You sure about change of semantics here? Inner `F` is supposed to be about releasing the old value.